Clean up LocalGcdHelper and docs

ajkannan opened the pull request:
* Adds an `options` method to `LocalGcdHelper` to make it more consistent with other test helpers.
* Uses the `options` method in `DatastoreTest`
* Updates the package info and TESTING docs to reflect the recently-added consistency setting
* Allows consistency to be set from command line invocation of `LocalGcdHelper`

ajkannan commented:
Will add a commit tomorrow to change the name of LocalGcdHelper to LocalDatastoreHelper to be more consistent with other libraries, as per @aozarov's suggestion.

aozarov commented on TESTING.md:
````diff
@@ -14,22 +14,22 @@ This library provides tools to help write tests for code that uses the following
+  - Run `LocalGcdHelper.java`'s `main` method with `START` provided as an argument, followed by optional arguments `--port=<port number>` and `--consistency=<float between 0 and 1, inclusive>`.  This will create a temporary folder on your computer and bind `localhost:<port number>` for communication with the local datastore.  If no port number is specified, port 8080 will be used.  The consistenc ...
+  - Use the `LocalGcdHelper.start(String projectId, int port, double consistency)` method before running your tests. For example, you can use the following code to start the local Datastore on any available port with the consistency set to 0.9:
+    ```java
+    int port = LocalGcdHelper.findAvailablePort(LocalGcdHelper.DEFAULT_PORT);
````
Any reason for not going all the way and automatically using port selection (and remove the static methods), so one should keep a reference to the same instance of the helper and stop it once tests complete.  At some point I was using the main method to start (and also calling from maven) but I think by now this is long gone, no?

mziccard commented on gcloud-java-datastore/src/main/java/com/google/gcloud/datastore/testing/LocalGcdHelper.java:
```diff
@@ -639,10 +653,13 @@ public static void main(String... args) throws IOException, InterruptedException
         parsedArgs.put("action", arg);
       } else {
-        throw new RuntimeException("Only accepts START, STOP, and --port=<port #> as arguments");
+        throw new RuntimeException("Only accepts START, STOP, --port=<port #> and "
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Avoid throwing raw exception types.](https://www.codacy.com/app/mziccard/gcloud-java/file/2325307737/issues/source?bid=3123103&fileBranchId=3205624#l661)

ajkannan commented:
Added a commit to:
* Rename LocalGcdHelper --> LocalDatastoreHelper and RemoteGcsHelper --> RemoteStorageHelper.
* Update TESTING docs, pkg-infos, and references to LocalGcdHelper/RemoteGcsHelper in javadoc.
* Remove main-related functions from LocalDatastoreHelper.

mziccard commented on gcloud-java-datastore/src/main/java/com/google/gcloud/datastore/testing/LocalDatastoreHelper.java:
```diff
@@ -101,14 +97,6 @@
-    }
-  }
-
   private static Path installedGcdPath() {
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [The method installedGcdPath() has an NPath complexity of 420](https://www.codacy.com/app/mziccard/gcloud-java/file/2329630392/issues/source?bid=3123103&fileBranchId=3205624#l100)

mziccard replied:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [The method installedGcdPath() has an NPath complexity of 420](https://www.codacy.com/app/mziccard/gcloud-java/file/2336495739/issues/source?bid=3123103&fileBranchId=3205624#l100)

mziccard commented on gcloud-java-storage/src/test/java/com/google/gcloud/storage/RemoteStorageHelperTest.java:
```diff
@@ -139,7 +139,7 @@ public void testForceDeleteTimeout() throws InterruptedException, ExecutionExcep
     EasyMock.expect(storageMock.delete(BUCKET_NAME)).andThrow(RETRYABLE_EXCEPTION).anyTimes();
     EasyMock.replay(storageMock);
-    assertTrue(!RemoteGcsHelper.forceDelete(storageMock, BUCKET_NAME, 50, TimeUnit.MICROSECONDS));
+    assertTrue(!RemoteStorageHelper.forceDelete(storageMock, BUCKET_NAME, 50, TimeUnit.MICROSECONDS));
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [assertTrue(!expr) can be replaced by assertFalse(expr)](https://www.codacy.com/app/mziccard/gcloud-java/file/2329630538/issues/source?bid=3123103&fileBranchId=3205624#l142)

mziccard replied:
Even though not related to this PR I think this comment is worth fixing

mziccard commented on TESTING.md:
````diff
@@ -11,55 +11,57 @@ This library provides tools to help write tests for code that uses the following
-  - Call `LocalGcdHelper.start(<project ID>, <port number>)` before running your tests.  Save the `LocalGcdHelper` object returned so that you can stop the emulator later.
+1. Start the local Datastore emulator before running your tests using `LocalDatastoreHelper`'s `start` method. This will create a temporary folder on your computer and bind a port for communication with the local datastore. There are two optional arguments for `start`: project ID and consistency. The consistency setting controls the fraction of Datastore writes that are immediately visible in glo ...
+  ```java
+  // Use defaults for project ID and consistency
````
Do you think it's worth mentioning what's the default consistency?

mziccard commented on TESTING.md:
````diff
@@ -11,55 +11,57 @@ This library provides tools to help write tests for code that uses the following
   ```java
-  DatastoreOptions options = DatastoreOptions.builder()
-      .projectId(PROJECT_ID)
+  DatastoreOptions options = LocalDatastoreHelper.options()
````
`LocalDatastoreHelper.options()` does not seem to exist. I think here we should use:
```java
DatastoreOptions.builder()
        .projectId("my-project-id")
        .host("http://<hostname of machine>:<port>")
        .authCredentials(AuthCredentials.noAuth())
        .build();
```

mziccard commented on gcloud-java-datastore/src/main/java/com/google/gcloud/datastore/testing/LocalDatastoreHelper.java:
```diff
@@ -589,80 +539,100 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
+   *     that setting this to 1.0 may mask incorrect assumptions about the consistency of
+   *     non-ancestor queries; non-ancestor queries are eventually consistent.
+   */
+  public static LocalDatastoreHelper start(String projectId, double consistency)
```
We could consider taking out the `findAndStartGcd()` call and renaming this (and the next) method to `create`. `findAndStartGcd()` could be renamed to `start()` and become public. This should be slightly more similar to remote helpers.

Also I think there might be a value in providing a way of setting the port if this class is ever used to start a remote emulator (I am thinking about firewalls here).

ajkannan replied:
On providing a way to set the port: after discussing with @aozarov, it seems like we should depend on the Google Cloud SDK to start up the emulator on a remote machine instead of using LocalGcdHelper.  As a result, providing a way to set the port would no longer be necessary.

It's most convenient to start the emulator on a remote machine via the main method (removed by this PR).  Before, we used the main method to start up gcd.sh for our own tests.  Since we don't use that anymore and there's an alternative (gcloud SDK), it seems like our LocalDatastoreHelper shouldn't really be used to start gcd on a remote machine.

Good point on the findAndStart --> create/start, will do.

ajkannan replied:
Ah, actually there's an issue with splitting up create and start that I hadn't considered until now.  When building DatastoreOptions, we make an RPC call when normalizing the input.  As a result, the user can't make these calls in the following order:

```java
LocalDatastoreHelper helper = LocalDatastoreHelper.create();
Datastore localDatastore = helper.options().service(); <-- will error
helper.start();
```

We can either document that you should only call options() after calling start(), but I think it's better to combine create and start in this case.

ajkannan replied:
Actually, I'll keep the create(), start(), and then document here and in other places that options() should be called after start(), as per @aozarov's suggestion.  This allows for starting/stopping the helper multiple times without having to create a new instance each time.

mziccard commented on gcloud-java-datastore/src/test/java/com/google/gcloud/datastore/DatastoreTest.java:
```diff
@@ -111,27 +110,19 @@
 
-  private static LocalGcdHelper gcdHelper;
-  private static final int PORT = LocalGcdHelper.findAvailablePort(LocalGcdHelper.DEFAULT_PORT);
+  private static LocalDatastoreHelper gcdHelper;
```
Should this variable be called `helper` or `datastoreHelper`?

mziccard commented on gcloud-java-datastore/src/test/java/com/google/gcloud/datastore/LocalDatastoreHelperTest.java:
```diff
@@ -0,0 +1,59 @@
+  private static final double TOLERANCE = 0.00001;
+
+  @Test
+  public void testStart() throws IOException, InterruptedException {
```
If we follow my suggestion of splitting `start` into `create` and `start` this test should be split and `start/stop` would no longer be needed to test that fields have been correctly initialized.

mziccard commented on gcloud-java-datastore/src/main/java/com/google/gcloud/datastore/testing/LocalDatastoreHelper.java:
```diff
@@ -589,80 +539,100 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
+   *     non-ancestor queries; non-ancestor queries are eventually consistent.
+   */
+  public static LocalDatastoreHelper start(String projectId, double consistency)
+      throws IOException, InterruptedException {
```
Sounds good to me!

ajkannan commented:
Added a commit to address comments, ready for another pass.

mziccard commented on TESTING.md:
````diff
@@ -13,19 +13,23 @@ This library provides tools to help write tests for code that uses the following
   Datastore localDatastore = helper.options().service();
   ```
   
+  Note that you must call `start` before `options()`; otherwise, you will see an `IOException` regarding a refused connection.
````
Should we intercept this situation and throw a more informative `IllegalStateException` with custom message: "Cannot get service options before starting the emulator"?

mziccard commented on TESTING.md:
````diff
@@ -38,14 +42,19 @@ You can test against a temporary local Datastore by following these steps:
+Note that the remote Datastore emulator must be running before running the code above. We recommend that you start the emulator on the remote machine using the [Google Cloud SDK](https://cloud.google.com/sdk/gcloud/reference/beta/emulators/datastore/) from command line, as shown below:
+
+```
+gcloud beta emulators datastore start
````
We should probably show how to set the port. If the emulator has to be remotely accessible I guess users would want it on a specific port.

```
gcloud beta emulators datastore start --host-port <hostname of machine>:<port>
```

mziccard commented on gcloud-java-datastore/src/main/java/com/google/gcloud/datastore/testing/LocalDatastoreHelper.java:
```diff
@@ -555,7 +555,7 @@ private static int findAvailablePort() {
   /**
    * Returns a {@link DatastoreOptions} instance that sets the host to use the Datastore emulator
-   * on localhost.
+   * on localhost. This should only be called after calling {@link #start}.
```
As I mentioned above I think here we should detected that the emulator has not started yet and we should throw an `IllegalStateException` with meaningful message.

mziccard replied:
Also we should document that exception

mziccard commented on gcloud-java-datastore/src/test/java/com/google/gcloud/datastore/LocalDatastoreHelperTest.java:
```diff
@@ -36,24 +36,22 @@
   public void testOptions() throws IOException, InterruptedException {
-    LocalDatastoreHelper helper = LocalDatastoreHelper.start();
+    LocalDatastoreHelper helper = LocalDatastoreHelper.create();
+    helper.start();
```
Any reason for removing `helper.stop()`?

ajkannan commented:
Rebased onto master and fixed all issues mentioned above.  At the suggestion of @aozarov, I now use a system property to determine whether the DatastoreOptions should issue and RPC to normalize the project ID.  As a result, the exception/ordering on start() and options() is no longer necessary.

mziccard commented on gcloud-java-datastore/src/main/java/com/google/gcloud/datastore/testing/LocalDatastoreHelper.java:
```diff
@@ -101,14 +99,6 @@
-    }
-  }
-
   private static Path installedGcdPath() {
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [The method installedGcdPath() has an NPath complexity of 420](https://www.codacy.com/app/mziccard/gcloud-java/file/2368544705/issues/source?bid=3123103&fileBranchId=3205624#l102)

mziccard replied:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [The method installedGcdPath() has an NPath complexity of 420](https://www.codacy.com/app/mziccard/gcloud-java/file/2368707250/issues/source?bid=3123103&fileBranchId=3205624#l102)

mziccard replied:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [The method installedGcdPath() has an NPath complexity of 420](https://www.codacy.com/app/mziccard/gcloud-java/file/2379270737/issues/source?bid=3123103&fileBranchId=3205624#l102)

aozarov commented on gcloud-java-datastore/src/main/java/com/google/gcloud/datastore/DatastoreOptions.java:
```diff
@@ -94,7 +95,8 @@ Builder normalizeDataset(boolean normalizeDataset) {
   private DatastoreOptions(Builder builder) {
     super(DatastoreFactory.class, DatastoreRpcFactory.class, builder);
-    normalizeDataset = builder.normalizeDataset;
+    String systemProjectId = System.getProperty(projectId());
```
I think we need to make it more specific.

e.g.
"com.google.gcloud.datastore.DatastoreOptions.skipNormalization.projectId", "true"
and then you can check it with [`Boolean.getBoolean(...)`](https://docs.oracle.com/javase/7/docs/api/java/lang/Boolean.html#getBoolean(java.lang.String)).

ajkannan commented:
Added a commit to use a more specific system property key.

mziccard commented on gcloud-java-datastore/src/main/java/com/google/gcloud/datastore/testing/LocalDatastoreHelper.java:
```diff
@@ -101,14 +100,6 @@
-    }
-  }
-
   private static Path installedGcdPath() {
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [The method installedGcdPath() has an NPath complexity of 420](https://www.codacy.com/app/mziccard/gcloud-java/file/2372680376/issues/source?bid=3123103&fileBranchId=3205624#l103)

ajkannan commented:
Added a commit to make `DatastoreOptions.Builder.normalizeDataset(boolean)` public rather than use system property because it's less magical.

ajkannan commented:
I had to rebase onto master because of a merge conflict.  Ready for another look.

aozarov commented on gcloud-java-datastore/src/main/java/com/google/gcloud/datastore/testing/LocalDatastoreHelper.java:
```diff
@@ -65,28 +63,27 @@
   private static final String GCD_BASENAME = "gcd-" + GCD_VERSION + "-" + GCD_BUILD;
   private static final String GCD_FILENAME = GCD_BASENAME + ".zip";
-  private static final String MD5_CHECKSUM = "72156cc993835c57f72789519b85249b";
+  private static final String MD5_CHECKSUM = "df876ba8f054d69acff30ec9540ec386";
```
was the gcd tool updated?

ajkannan replied:
Yes, a newer version was made available, so I figured I'd use this cleanup PR as an opportunity to update our link.

aozarov commented on gcloud-java-datastore/src/main/java/com/google/gcloud/datastore/testing/LocalDatastoreHelper.java:
```diff
@@ -589,80 +540,102 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
-        return;
-      default:
-        break;
+  private static int findAvailablePort() {
```
Let's create an issue for creating a base class for test helpers.
This is an example of code that could be shared.

ajkannan replied:
Issue #879 created.

aozarov commented on gcloud-java-datastore/src/main/java/com/google/gcloud/datastore/testing/LocalDatastoreHelper.java:
```diff
@@ -65,28 +63,27 @@
   private static final Path INSTALLED_GCD_PATH;
   private static final String GCD_VERSION_PREFIX = "gcd-emulator ";
-  private static final double DEFAULT_CONSISTENCY = 0.9;
+  private static final String PROJECT_ID_PREFIX = "test-id-";
```
maybe "test_project_"

ajkannan replied:
I can change this to "test-project-" (dashes instead of underscores because the client checks for a valid project ID, and underscores aren't allowed).

aozarov commented on gcloud-java-datastore/src/main/java/com/google/gcloud/datastore/testing/LocalDatastoreHelper.java:
```diff
@@ -589,80 +540,102 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
+   */
+  public void start() throws IOException, InterruptedException {
+    // send a quick request in case we have a hanging process from a previous run
+    checkArgument(consistency >= 0.0 && consistency <= 1.0, "Consistency must be between 0 and 1");
```
This could move to the constructor.

aozarov commented on gcloud-java-datastore/src/main/java/com/google/gcloud/datastore/testing/LocalDatastoreHelper.java:
```diff
@@ -589,80 +540,102 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
+  /**
+   * Returns the port on localhost to which the local Datastore listens for requests.
+   */
+  public int port() {
```
I am not sure this is necessary. Do you?

ajkannan replied:
I don't see a strong use case for it, I can remove it.

aozarov commented on gcloud-java-datastore/src/test/java/com/google/gcloud/datastore/DatastoreTest.java:
```diff
@@ -68,7 +67,11 @@
-  private static final String PROJECT_ID = LocalGcdHelper.DEFAULT_PROJECT_ID;
+  private static LocalDatastoreHelper helper = LocalDatastoreHelper.create(1.0);
+  private static final DatastoreOptions options =
+      helper.options().toBuilder().retryParams(RetryParams.noRetries()).build();
```
Any reason for not using `RetryParams.noRetries()` by default for the local service helper?

ajkannan replied:
No reason, will change.

aozarov commented on gcloud-java-datastore/src/test/java/com/google/gcloud/datastore/DatastoreTest.java:
```diff
@@ -162,8 +151,8 @@ public void setUp() {
   public static void afterClass() throws IOException, InterruptedException {
-    if (gcdHelper != null) {
-      gcdHelper.stop();
+    if (helper != null) {
```
I am not sure this check is necessary (and it is not consistent with beforeClass).

aozarov commented:
Looks good. Few comments. Feel free to merge after considering them.

aozarov commented on gcloud-java-datastore/src/test/java/com/google/gcloud/datastore/LocalDatastoreHelperTest.java:
```diff
@@ -0,0 +1,53 @@
+public class LocalDatastoreHelperTest {
+
+  private static final double TOLERANCE = 0.00001;
+  private static final String PROJECT_ID_PREFIX = "test-project-";
```
Just realized that this test class does not match the same package structure as the class it tests.
I think we should change that. See #880 
And when we do so, you can change `LocalDatastoreTestHelper.PROJECT_ID_PREFIX` to be package scope and use it here.
